services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/extensions/DeleteBucketAndAllContents.java

```diff
@@ -0,0 +1,138 @@
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.services.s3.internal.extensions;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.stream.Stream;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.extensions.S3ClientSdkExtension;
+import software.amazon.awssdk.services.s3.model.DeleteMarkerEntry;
+import software.amazon.awssdk.services.s3.model.DeleteObjectsResponse;
+import software.amazon.awssdk.services.s3.model.ListObjectVersionsResponse;
+import software.amazon.awssdk.services.s3.model.ObjectIdentifier;
+import software.amazon.awssdk.services.s3.model.ObjectVersion;
+import software.amazon.awssdk.services.s3.model.S3Object;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.Validate;
+
+@SdkInternalApi
+public class DeleteBucketAndAllContents implements S3ClientSdkExtension {
+    private static final Logger log = Logger.loggerFor(DeleteBucketAndAllContents.class);
+    private static final int MAX_DELETE_OBJECTS_SIZE = 1_000;
+
+    private final S3Client s3;
+
+    public DeleteBucketAndAllContents(S3Client s3) {
+        this.s3 = Validate.notNull(s3, "s3");
+    }
+
+    @Override
+    public void deleteBucketAndAllContents(String bucket) {
+        Validate.notEmpty(bucket, "bucket");
+        log.debug(() -> "Emptying bucket: " + bucket);
+        deleteAllObjects(bucket);
+        deleteAllVersions(bucket);
+        log.debug(() -> "Deleting bucket: " + bucket);
+        s3.deleteBucket(r -> r.bucket(bucket));
+    }
+
+    private void deleteAllObjects(String bucket) {
+        Stream<ObjectIdentifier> stream = s3.listObjectsV2Paginator(r -> r.bucket(bucket))
+                                            .contents()
+                                            .stream()
+                                            .map(this::toDeletePojo);
+        deleteObjectStream(bucket, stream);
+    }
+
+    private void deleteAllVersions(String bucket) {
+        Stream<ObjectIdentifier> stream = s3.listObjectVersionsPaginator(r -> r.bucket(bucket))
+                                            .stream()
+                                            .map(this::toDeletePojos)
+                                            .flatMap(Collection::stream);
+        deleteObjectStream(bucket, stream);
+    }
+
+    private void deleteObjectStream(String bucket, Stream<ObjectIdentifier> stream) {
+        Iterator<List<ObjectIdentifier>> batchIterator = batch(stream.iterator(), MAX_DELETE_OBJECTS_SIZE);
+        batchIterator.forEachRemaining(objects -> {
+            log.debug(() -> String.format("Deleting %s objects: %s", objects.size(), objects));
```

Comment:
What does this line look like in the log, specifically `objects`?

Reply:
It logs the key names, which may not be appropriate. I was considering removing the logs altogether for that reason.

Here's an example from `CopySourceIntegrationTest`:

```
Deleting S3 bucket: copy-source-integ-test-src-lyncbenn-2036
10:18:10.213 [main] DEBUG software.amazon.awssdk.services.s3.internal.extensions.DeleteBucketAndAllContents - Emptying bucket: copy-source-integ-test-src-lyncbenn-2036
10:18:10.398 [main] DEBUG software.amazon.awssdk.services.s3.internal.extensions.DeleteBucketAndAllContents - Deleting 5 objects: [ObjectIdentifier(Key=%20), ObjectIdentifier(Key=key/with/slashes), ObjectIdentifier(Key=simpleKey), ObjectIdentifier(Key=specialChars/ +!#$&'()*,:;=?@"), ObjectIdentifier(Key=🪣)]
10:18:10.943 [main] DEBUG software.amazon.awssdk.services.s3.internal.extensions.DeleteBucketAndAllContents - Deleting bucket: copy-source-integ-test-src-lyncbenn-2036
Deleting S3 bucket: copy-source-integ-test-versioned-src-lyncbenn-1687
10:18:11.235 [main] DEBUG software.amazon.awssdk.services.s3.internal.extensions.DeleteBucketAndAllContents - Emptying bucket: copy-source-integ-test-versioned-src-lyncbenn-1687
10:18:11.335 [main] DEBUG software.amazon.awssdk.services.s3.internal.extensions.DeleteBucketAndAllContents - Deleting 5 objects: [ObjectIdentifier(Key=%20), ObjectIdentifier(Key=key/with/slashes), ObjectIdentifier(Key=simpleKey), ObjectIdentifier(Key=specialChars/ +!#$&'()*,:;=?@"), ObjectIdentifier(Key=🪣)]
10:18:11.909 [main] DEBUG software.amazon.awssdk.services.s3.internal.extensions.DeleteBucketAndAllContents - Deleting 20 objects: [ObjectIdentifier(Key=%20, VersionId=V.sVzY.75kBdHEqJM39by_Z5sTBHZ7fQ), ObjectIdentifier(Key=%20, VersionId=sWggSroHE_WHRIOSZaDKZ2brqRu7yZhO), ObjectIdentifier(Key=%20, VersionId=fhPc80IYqCW9CjP.STRJNF2B1vlj4wHB), ObjectIdentifier(Key=key/with/slashes, VersionId=Gav4gvn8A5Ip.Ofmu8w0qmDyZWeFOI_T), ObjectIdentifier(Key=key/with/slashes, VersionId=7AhHansYt7WA5ON1ZEJ0EE0QREBE4.Pk), ObjectIdentifier(Key=key/with/slashes, VersionId=NsKZD8LYIBXs_YECP54aNSJ0QDgFgbYm), ObjectIdentifier(Key=simpleKey, VersionId=H5nEdAB_uTfl7ZA05WG0875hlfpoRn.y), ObjectIdentifier(Key=simpleKey, VersionId=doUGPN_k40B.sh8mGSGooGFn4TKvW1XE), ObjectIdentifier(Key=simpleKey, VersionId=3isxbpv5bC4u0SWKP_32YZ_h7d7VW_Y4), ObjectIdentifier(Key=specialChars/ +!#$&'()*,:;=?@", VersionId=dOYfzJl.xJe.ipWuuilM1M3jtLSYU7Dz), ObjectIdentifier(Key=specialChars/ +!#$&'()*,:;=?@", VersionId=duQuW7pjfUxRim5080eXlAZ83C8CfbRk), ObjectIdentifier(Key=specialChars/ +!#$&'()*,:;=?@", VersionId=VjBGda03yhKoVHCkz6c6DyiqXk3XAZH3), ObjectIdentifier(Key=🪣, VersionId=2IDkpsK3YOGXjvPefVkEOVk0aj7_clYf), ObjectIdentifier(Key=🪣, VersionId=IA2lPaG1Kf57vtCv2ebcVA_h12aHVdLC), ObjectIdentifier(Key=🪣, VersionId=RcMPfGgo0ejdLy_mlDXiHP5w1B9nsR_d), ObjectIdentifier(Key=%20, VersionId=Q25OFLzsmtzUIiU7cIDEZ2AWBH1IgQSJ), ObjectIdentifier(Key=key/with/slashes, VersionId=4bUxcv.SPPW0WwkejxKU9X2PnAz1XkRO), ObjectIdentifier(Key=simpleKey, VersionId=kUHpkEd3zuL6k2kkXfSe5QYTHlGPaKnE), ObjectIdentifier(Key=specialChars/ +!#$&'()*,:;=?@", VersionId=xof9bJbMxbR7WhXsQ8wGFHfn.1rWuioL), ObjectIdentifier(Key=🪣, VersionId=ZfRQ8UITh3_qflMbXHOCV9nRKP9AKcMK)]
10:18:12.300 [main] DEBUG software.amazon.awssdk.services.s3.internal.extensions.DeleteBucketAndAllContents - Deleting bucket: copy-source-integ-test-versioned-src-lyncbenn-1687
Deleting S3 bucket: copy-source-integ-test-dest-lyncbenn-7871
10:18:12.830 [main] DEBUG software.amazon.awssdk.services.s3.internal.extensions.DeleteBucketAndAllContents - Emptying bucket: copy-source-integ-test-dest-lyncbenn-7871
10:18:12.915 [main] DEBUG software.amazon.awssdk.services.s3.internal.extensions.DeleteBucketAndAllContents - Deleting 5 objects: [ObjectIdentifier(Key=%20), ObjectIdentifier(Key=key/with/slashes), ObjectIdentifier(Key=simpleKey), ObjectIdentifier(Key=specialChars/ +!#$&'()*,:;=?@"), ObjectIdentifier(Key=🪣)]
10:18:13.653 [main] DEBUG software.amazon.awssdk.services.s3.internal.extensions.DeleteBucketAndAllContents - Deleting bucket: copy-source-integ-test-dest-lyncbenn-7871
```

Reply:
Ok, agreed to remove user info.